src/emulator/pubsubEmulator.ts

```diff
@@ -105,20 +120,71 @@ export class PubsubEmulator implements EmulatorInstance {
       this.onMessage(topicName, message);
     });
 
-    topicTriggers.add(trigger);
-    this.triggers.set(topicName, topicTriggers);
-    this.subscriptions.set(topicName, sub);
+    triggers.push({ triggerKey, signatureType });
+    this.triggersForTopic.set(topicName, triggers);
+    this.subscriptionForTopic.set(topicName, sub);
+  }
+
+  private getRequestOptions(
+    topic: string,
+    message: Message,
+    signatureType: SignatureType
+  ): Record<string, unknown> {
+    const baseOpts = {
+      origin: `http://${EmulatorRegistry.getInfoHostString(
+        EmulatorRegistry.get(Emulators.FUNCTIONS)!.getInfo()
+      )}`,
+    };
+    if (signatureType === "event") {
+      return {
+        ...baseOpts,
+        data: {
+          context: {
+            eventId: uuid.v4(),
+            resource: {
+              service: "pubsub.googleapis.com",
+              name: `projects/${this.args.projectId}/topics/${topic}`,
+            },
+            eventType: "google.pubsub.topic.publish",
+            timestamp: message.publishTime.toISOString(),
+          },
+          data: {
+            data: message.data,
+            attributes: message.attributes,
+          },
+        },
+      };
+    } else if (signatureType === "cloudevent") {
+      const data: MessagePublishedData = {
+        message: {
+          messageId: message.id,
+          publishTime: message.publishTime,
+          attributes: message.attributes,
+          orderingKey: message.orderingKey,
+          data: message.data.toString("base64"),
+        },
+        subscription: this.subscriptionForTopic.get(topic)!.name,
+      };
+      const ce = HTTP.structured(
+        new CloudEvent({
+          type: "google.cloud.pubsub.topic.v1.messagePublished",
+          source: `//pubsub.googleapis.com/projects/${this.args.projectId}/topics/${topic}`,
+          data,
+        })
+      );
+      return { ...baseOpts, data: ce.body, headers: ce.headers };
```

Comment:
You don't really need to worry about ce.body and ce.headers if you don't want to. You can just set the contenttype to the cloudevent content type and put everything in body (structured vs binary encoding; both are allowed in the spec)

Reply:
Done!